backend/crawler/crawler.py

```diff
@@ -153,43 +219,71 @@ def parse(year, semester, lectures):
-                                                college=college_instance,
-                                                department=department_instance,
-                                                # major=
-                                                )
-        course_instance.save()
+                department_instance.save()
+
+            else:
+                department_instance = None
+
+        try:
+            major_instance = Major.objects.get(name=lecture['major'])
+        except Exception as e:
+            if lecture['major'] != '':
+                major_instance = Major.objects.create(department=department_instance,
+                                                  name=lecture['major'])
+                major_instance.save()
+
+            else:
+                major_instance = None
+
+        try:
+            course_instance = Course.objects.get(name=lecture['name'])
+
+        except Exception as e:
+            course_instance = Course.objects.create(code=lecture['code'],
+                                                    name=lecture['name'],
+                                                    type=lecture['type'],
+                                                    # field=,
+                                                    grade=int(lecture['grade'][0]),
+                                                    credit=int(lecture['credit'].split('-')[0]),
+                                                    college=college_instance,
+                                                    department=department_instance,
+                                                    major=major_instance,
+                                                    )
+            course_instance.save()
 
         lecture_instance = Lecture.objects.create(course=course_instance,
                                                   year=year,
                                                   semester=semester,
                                                   number=lecture['number'],
                                                   instructor=lecture['instructor'],
-                                                  note='')
+                                                  note=lecture['note'])
 
         # It really is absurd, but there exists lectures without any time slot.
         try:
             for time_slot in lecture['time_slots']:
-                classroom_instance = Classroom.objects.create(building=time_slot['classroom']['building'],
-                                                              room_no=time_slot['classroom']['room_no'])
-                classroom_instance.save()
+                building = time_slot['classroom']['building']
+                room_no = time_slot['classroom']['room_no']
+                whole = building + '-' + room_no
+                try:
+                    classroom_instance = Classroom.objects.get(whole=whole)
```

Comment:
I think the `whole` field is redundant because we can yield same result like this,
```python3
classroom_instance = Classroom.objects.get(building=building, room_no=room_no)
```
And, to set the uniqueness for the model, please refer to this section [unique_together](https://docs.djangoproject.com/en/dev/ref/models/options/#unique-together)
But, since I have never used this option, I cannot gaurantee the result you want.